ui/src/chat/ChatMessage/ChatMessageOptions.tsx

```diff
@@ -102,91 +117,171 @@ export default function ChatMessageOptions(props: {
     }
   }, [isMobile, loadEmoji]);
 
+  const showReactAction = canWrite;
+  const showReplyAction = !hideReply;
+  const showThreadAction =
+    !writ.memo.replying && writ.memo.replying?.length !== 0 && !hideThreadReply;
+  const showCopyAction = !!groupFlag;
+  const showDeleteAction = isAdmin || window.our === writ.memo.author;
+
+  const actions: Action[] = [];
+
+  if (showReactAction) {
+    actions.push({
+      key: 'react',
+      content: (
+        <div className="flex items-center" aria-label="React">
+          <FaceIcon className="mr-2 h-6 w-6" />
+          React
+        </div>
+      ),
+      onClick: () => {
+        navigate(`picker/${writ.seal.id}`, {
+          state: { backgroundLocation: location },
+        });
+      },
+    });
+  }
+
+  if (showReplyAction) {
+    actions.push({
+      key: 'reply',
+      content: (
+        <div className="flex items-center">
+          <BubbleIcon className="mr-2 h-6 w-6" />
+          Reply
+        </div>
+      ),
+      onClick: reply,
+    });
+  }
+
+  if (showThreadAction) {
+    actions.push({
+      key: 'thread',
+      content: (
+        <div className="flex items-center">
+          <HashIcon className="mr-2 h-6 w-6" />
+          Start Thread
+        </div>
+      ),
+      onClick: () => navigate(`message/${writ.seal.id}`),
+    });
+  }
+
+  if (showCopyAction) {
+    actions.push({
+      key: 'copy',
+      content: (
+        <div className="flex items-center">
+          {didCopy ? (
+            <CheckIcon className="mr-2 h-6 w-6" />
+          ) : (
+            <CopyIcon className="mr-2 h-6 w-6" />
+          )}
+          {didCopy ? 'Copied!' : 'Copy'}
+        </div>
+      ),
+      onClick: onCopy,
+      keepOpenOnClick: true,
+    });
+  }
+
+  if (showDeleteAction) {
+    actions.push({
+      key: 'delete',
+      type: 'destructive',
+      content: (
+        <div className="flex items-center">
+          <XIcon className="mr-2 h-6 w-6" />
+          Delete
+        </div>
+      ),
+      onClick: () => setDeleteOpen(true),
+      keepOpenOnClick: true,
+    });
+  }
+
+  if (!open && !isMobile) {
+    return null;
+  }
+
   return (
-    <div className="absolute right-2 -top-5 z-10 h-full">
-      <div
-        data-testid="chat-message-options"
-        className="sticky top-0 flex space-x-0.5 rounded-lg border border-gray-100 bg-white p-[1px] align-middle"
-      >
-        {canWrite && !isMobile ? (
-          <EmojiPicker
-            open={pickerOpen}
-            setOpen={setPickerOpen}
-            onEmojiSelect={onEmoji}
-            withTrigger={false}
+    <>
+      {isMobile ? (
+        <ActionMenu open={open} onOpenChange={onOpenChange} actions={actions} />
+      ) : (
```

Comment:
We overloaded what "open" means here. Originally we had logic to not render the options at all (one level above in `ChatMessage`) because it was expensive to render all of the radix-ui components underneath since there were so many messages. 

I think you still want to not render anything here if `open === false` (which maybe should be renamed). And then if it is "open" then action menu should just be passed `true` for it's open prop.

Reply:
Totally understand you, but let me explain how it works:
- For desktop - we use the same logic as before (don't render at all if `!open`)
- For mobile, on contrary, we DO want actually the component to be rendered, because it uses the animation for dismissing (otherwise it's disappearing immediately, which is not so great).

But still, you don't need to be afraid that it renders for every message, cause under the hood when it's not open the dialog doesn't add anything to the DOM tree, only when it's open it adds the dialog component to the body container.
You can prove it to yourself by looking at the Elements tab.

So, no performance regression 🫡